tools/traffic/workers/blob_reader.go

```diff
@@ -0,0 +1,231 @@
+// Start begins a blob reader goroutine.
+func (r *BlobReader) Start() {
+	r.waitGroup.Add(1)
+	ticker := time.NewTicker(r.config.ReadRequestInterval)
+	go func() {
+		defer r.waitGroup.Done()
+		for {
+			select {
+			case <-(*r.ctx).Done():
+				err := (*r.ctx).Err()
+				if err != nil {
+					r.logger.Info("blob r context closed", "err:", err)
+				}
+				return
+			case <-ticker.C:
+				r.randomRead()
+			}
+		}
+	}()
+}
+
+// randomRead reads a random blob.
+func (r *BlobReader) randomRead() {
+	metadata := r.blobsToRead.GetNext()
+	if metadata == nil {
+		// There are no blobs that we are required to read.
+		return
+	}
+
+	r.metrics.requiredReadPoolSizeMetric.Set(float64(r.blobsToRead.Size()))
+
+	ctxTimeout, cancel := context.WithTimeout(*r.ctx, r.config.FetchBatchHeaderTimeout)
+	defer cancel()
+	batchHeader, err := metrics.InvokeAndReportLatency(r.metrics.fetchBatchHeaderMetric,
+		func() (*contractEigenDAServiceManager.IEigenDAServiceManagerBatchHeader, error) {
+			return r.chainClient.FetchBatchHeader(
+				ctxTimeout,
+				gcommon.HexToAddress(r.config.EigenDAServiceManager),
+				metadata.BatchHeaderHash[:],
+				big.NewInt(int64(0)),
+				nil)
+		})
+	if err != nil {
+		r.logger.Error("failed to get batch header", "err:", err)
+		r.metrics.fetchBatchHeaderFailure.Increment()
+		return
+	}
+	r.metrics.fetchBatchHeaderSuccess.Increment()
+
+	ctxTimeout, cancel = context.WithTimeout(*r.ctx, r.config.RetrieveBlobChunksTimeout)
+	chunks, err := metrics.InvokeAndReportLatency(r.metrics.readLatencyMetric, func() (*clients.BlobChunks, error) {
+		return r.retriever.RetrieveBlobChunks(
+			ctxTimeout,
+			metadata.BatchHeaderHash,
+			uint32(metadata.BlobIndex),
+			uint(batchHeader.ReferenceBlockNumber),
+			batchHeader.BlobHeadersRoot,
+			core.QuorumID(0))
+	})
+	cancel()
+	if err != nil {
+		r.logger.Error("failed to read chunks", "err:", err)
+		r.metrics.readFailureMetric.Increment()
+		return
+	}
+	r.metrics.readSuccessMetric.Increment()
+
+	assignments := chunks.Assignments
+
+	data, err := r.retriever.CombineChunks(chunks)
+	if err != nil {
+		r.logger.Error("failed to combine chunks", "err:", err)
+		r.metrics.recombinationFailureMetric.Increment()
+		return
+	}
+	r.metrics.recombinationSuccessMetric.Increment()
+
+	r.verifyBlob(metadata, &data)
+
+	indexSet := make(map[encoding.ChunkNumber]bool)
+	for index := range chunks.Indices {
+		indexSet[chunks.Indices[index]] = true
+	}
+
+	for id, assignment := range assignments {
+		for index := assignment.StartIndex; index < assignment.StartIndex+assignment.NumChunks; index++ {
+			if indexSet[index] {
+				r.reportChunk(id)
+			} else {
+				r.reportMissingChunk(id)
+			}
+		}
+	}
+}
+
+// reportChunk reports a successful chunk read.
+func (r *BlobReader) reportChunk(operatorId core.OperatorID) {
+	metric, exists := r.metrics.operatorSuccessMetrics[operatorId]
+	if !exists {
+		metric = r.metrics.generatorMetrics.NewCountMetric(fmt.Sprintf("operator_%x_returned_chunk", operatorId))
```

Comment:
So there will be hundreds of metrics, not sure if that's a quite usable thing.

Reply:
I agree that this is going to be noisy. Is it critical for this utility to determine which nodes are failing to return chunks, or would it be ok just to report the fraction of nodes that return chunks?

If knowledge of which nodes are failing to return chunks is important, do you have any recommendations for how to extract this sort of data without using metrics this way?

Reply:
Hmm would keeping track of counts for only failed retrieval reduce the number of metrics here?

Reply:
I think this could be updated in this PR

Reply:
This part needs to be careful about cardinality. SG to follow up to cap the scope here. LMK if you need some inputs on refactoring this later.

Reply:
The primary reason why I'd like to address metrics as a follow up is because there are other metrics I added in prior PRs that also need to be considered. Makes sense to me to do all the metrics at once.